exchanges/coinbasepro/coinbasepro.go

```diff
@@ -3,761 +3,1283 @@ package coinbasepro
+func (c *CoinbasePro) GetHistoricRates(ctx context.Context, productID, granularity string, startDate, endDate time.Time, authenticated bool) ([]CandleStruct, error) {
+	if productID == "" {
+		return nil, errProductIDEmpty
+	}
+	validGran := common.StringSliceContains(allowedGranularities, granularity)
+	if !validGran {
+		return nil, fmt.Errorf("%w %v, allowed granularities are: %+v", kline.ErrUnsupportedInterval, granularity, allowedGranularities)
+	}
+	vals := url.Values{}
+	vals.Set("start", strconv.FormatInt(startDate.Unix(), 10))
+	vals.Set("end", strconv.FormatInt(endDate.Unix(), 10))
+	vals.Set("granularity", granularity)
+	resp := struct {
+		Candles []CandleStruct `json:"candles"`
+	}{}
+	if authenticated {
+		path := coinbaseV3 + coinbaseProducts + "/" + productID + "/" + coinbaseCandles
+		return resp.Candles, c.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, path, vals, nil, true, &resp, nil)
+	}
+	path := coinbaseV3 + coinbaseMarket + "/" + coinbaseProducts + "/" + productID + "/" + coinbaseCandles
+	return resp.Candles, c.SendHTTPRequest(ctx, exchange.RestSpot, path, vals, &resp)
 }
 
-// GetHistoricRates returns historic rates for a product. Rates are returned in
-// grouped buckets based on requested granularity.
-func (c *CoinbasePro) GetHistoricRates(ctx context.Context, currencyPair, start, end string, granularity int64) ([]History, error) {
-	values := url.Values{}
+// GetTicker returns snapshot information about the last trades (ticks) and best bid/ask. Contrary to documentation, this does not tell you the 24h volume
+func (c *CoinbasePro) GetTicker(ctx context.Context, productID string, limit uint16, startDate, endDate time.Time, authenticated bool) (*Ticker, error) {
+	if productID == "" {
+		return nil, errProductIDEmpty
+	}
+	vals := url.Values{}
+	vals.Set("limit", strconv.FormatInt(int64(limit), 10))
+	if !startDate.IsZero() && !startDate.Equal(time.Time{}) {
+		vals.Set("start", strconv.FormatInt(startDate.Unix(), 10))
+	}
+	if !endDate.IsZero() && !endDate.Equal(time.Time{}) {
+		vals.Set("end", strconv.FormatInt(endDate.Unix(), 10))
+	}
+	var resp Ticker
+	if authenticated {
+		path := coinbaseV3 + coinbaseProducts + "/" + productID + "/" + coinbaseTicker
+		return &resp, c.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, path, vals, nil, true, &resp, nil)
+	}
+	path := coinbaseV3 + coinbaseMarket + "/" + coinbaseProducts + "/" + productID + "/" + coinbaseTicker
+	return &resp, c.SendHTTPRequest(ctx, exchange.RestSpot, path, vals, &resp)
+}
 
-	if start != "" {
-		values.Set("start", start)
-	} else {
-		values.Set("start", "")
+// PlaceOrder places either a limit, market, or stop order
+func (c *CoinbasePro) PlaceOrder(ctx context.Context, clientOID, productID, side, stopDirection, orderType, stpID, marginType, rpID string, amount, limitPrice, stopPrice, leverage float64, postOnly bool, endTime time.Time) (*PlaceOrderResp, error) {
```

Comment:
Suggestion: There is so much going on here for params I would change it to a struct it shouldn't be an issue in terms of allocation because it shouldn't escape scope.